### PR TITLE
Raise a KeyboardInterrupt on a SIGINT

### PR DESCRIPTION
**Pull request by coolreader18**
This is based on #1311 for the PySyscallFunctions struct.

### CODE REVIEW

File: vm/src/stdlib/signal.rs [hunk 1]
```diff
@@ -23,51 +23,64 @@ const SIG_IGN: libc::sighandler_t = 1;
...
 
 // We cannot use the NSIG const in the arr macro. This will fail compilation if NSIG is different.
-static mut TRIGGERS: [AtomicBool; NSIG] = arr![AtomicBool::new(false); 64];
+static TRIGGERS: [AtomicBool; NSIG] = arr![AtomicBool::new(false); 64];
 
 extern "C" fn run_signal(signum: i32) {
-    unsafe {
-        TRIGGERS[signum as usize].store(true, Ordering::Relaxed);
-    }
+    TRIGGERS[signum as usize].store(true, Ordering::Relaxed);
```
Comment (windelbouwman) [inline vm/src/stdlib/signal.rs]: Nice one! The `AtomicBool` appears the be mutable itself!

File: vm/src/stdlib/signal.rs [hunk 2]
```diff
@@ -23,51 +23,64 @@ const SIG_IGN: libc::sighandler_t = 1;
...
+    );
     Ok(old_handler)
 }
 
-fn getsignal(signalnum: PyIntRef, vm: &VirtualMachine) -> PyResult<Option<PyObjectRef>> {
-    let signalnum = signalnum.as_bigint().to_i32().unwrap();
-    Ok(vm.signal_handlers.borrow_mut().get(&signalnum).cloned())
+fn getsignal(signalnum: i32, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+    assert_in_range(signalnum, vm)?;
+    Ok(vm.signal_handlers.borrow()[signalnum as usize].clone())
```
Comment (windelbouwman) [inline vm/src/stdlib/signal.rs]: We should probably check for signalnum boundaries in the array.

File: vm/src/stdlib/time_module.rs [hunk 3]
```diff
@@ -22,36 +22,62 @@ use num_traits::cast::ToPrimitive;
...
+}
+
+fn time_sleep(seconds: PyFloatRef, vm: &VirtualMachine) -> PyResult<()> {
+    let seconds = seconds.to_f64();
     let secs: u64 = seconds.trunc() as u64;
     let nanos: u32 = (seconds.fract() * 1e9) as u32;
     let duration = Duration::new(secs, nanos);
-    thread::sleep(duration);
-    Ok(vm.get_none())
+    let sleep_func = vm.settings.syscall_functions.sleep.unwrap_or(do_sleep);
```
Comment (windelbouwman) [inline vm/src/stdlib/time_module.rs]: See my other comment in PR #1311 about this construction. I think we should not add the `syscall_functions` member.

File: vm/src/stdlib/signal.rs (hunk 2 above)
Comment (coolreader18) [reply vm/src/stdlib/signal.rs]: That's what `assert_in_range` does.

File: vm/src/stdlib/time_module.rs (hunk 3 above)
Comment (coolreader18) [reply vm/src/stdlib/time_module.rs]: I decoupled this PR from that change, we can look into that separately.

File: vm/src/stdlib/signal.rs (hunk 2 above)
Comment (coolreader18) [reply vm/src/stdlib/signal.rs]: That's what `assert_in_range` does.

File: vm/src/vm.rs [hunk 4]
```diff
@@ -166,7 +168,10 @@ impl VirtualMachine {
         let import_func = RefCell::new(ctx.none());
         let profile_func = RefCell::new(ctx.none());
         let trace_func = RefCell::new(ctx.none());
+
         let vm = VirtualMachine {
+            #[cfg(not(target_arch = "wasm32"))]
```
Comment (windelbouwman) [inline vm/src/vm.rs]: We can have this signal handlers in wasm as well right? They do not hurt us there, so this `#[cfg]` could be removed.

File: src/main.rs [hunk 5]
```diff
@@ -486,12 +486,10 @@ fn get_history_path() -> PathBuf {
     xdg_dirs.place_cache_file("repl_history.txt").unwrap()
 }
 
-fn get_prompt(vm: &VirtualMachine, prompt_name: &str) -> String {
```
Comment (windelbouwman) [inline src/main.rs]: I would like to keep the old signature and return a `String` here. What is the benefit of returning the `Option` here?

File: vm/src/stdlib/signal.rs (hunk 1 above)
Comment (windelbouwman) [reply vm/src/stdlib/signal.rs]: Btw, did you run benchmarks on this change? This change might trigger a performance boost :)

File: vm/src/stdlib/signal.rs (hunk 2 above)
Comment (windelbouwman) [reply vm/src/stdlib/signal.rs]: Ah, I missed that. Thanks!

File: vm/src/vm.rs (hunk 4 above)
Comment (coolreader18) [reply vm/src/vm.rs]: It uses the `NSIG` constant from `signal.rs`; should I move that into `vm.rs`?

File: src/main.rs (hunk 5 above)
Comment (coolreader18) [reply src/main.rs]: It doesn't clone the string this way, `vm.to_str` returns a `PyStringRef` whose inner `String` is already held inside of a PyObjectRef on the heap, but `vm.to_pystr` clones that `String` when all `repl.readline()` needs is an `&str`. I know it's a really tiny micro-optimization, especially when 99% of the time the string's only going to be 5 characters long, but I just like doing them.

File: vm/src/vm.rs (hunk 4 above)
Comment (windelbouwman) [reply vm/src/vm.rs]: Yes, this is probably a good idea to keep the `NSIG` constant where the array is defined / created.

File: src/main.rs [hunk 6]
```diff
@@ -542,18 +541,18 @@ fn run_shell(vm: &VirtualMachine, scope: Scope) -> PyResult<()> {
...
                     }
                 }
             }
             Err(ReadlineError::Interrupted) => {
-                // TODO: Raise a real KeyboardInterrupt exception
-                println!("^C");
+                let exc = vm
+                    .new_empty_exception(vm.ctx.exceptions.keyboard_interrupt.clone())
+                    .unwrap();
+                print_exception(vm, &exc);
```
Comment (windelbouwman) [inline src/main.rs]: There is something which occurs strange to me when examining this piece of code. The exception is created and printed, but should it not be inserted into the virtualmachine in some way? Maybe we can inject it into the virtual machine in a similar manner in which the signals are inserted into the virtualmachine, with the check_signals function and some sort of a queue.

I would suggest to add a function to the virtualmachine, named something like `inject_exception`, which will put this exception object into a queue, or a slot, and the frame eval loop will pick it up.

**Comment by windelbouwman** [comment]
@palaviv could you have a look at this change as well?

File: src/main.rs (hunk 6 above)
Comment (coolreader18) [reply src/main.rs]: The problem here is that the code I was converting from CPython here is actually from the builtin `input()` function, which then gets used by the `code` Python module to run the shell. Would it be okay if I left it like this for now but added `input()` and `code.py` in a separate PR?

File: vm/src/stdlib/signal.rs [hunk 7]
```diff
@@ -23,51 +23,64 @@ const SIG_IGN: libc::sighandler_t = 1;
...
-    let sig_dfl = vm.get_attribute(signal_module.clone(), "SIG_DFL")?;
-    let sig_ign = vm.get_attribute(signal_module, "SIG_IGN")?;
-    let signalnum = signalnum.as_bigint().to_i32().unwrap();
-    check_signals(vm);
-    let sig_handler = if handler.is(&sig_dfl) {
-        SIG_DFL
-    } else if handler.is(&sig_ign) {
-        SIG_IGN
+fn assert_in_range(signum: i32, vm: &VirtualMachine) -> PyResult<()> {
+    if (1..=NSIG as i32).contains(&signum) {
```
Comment (palaviv) [inline vm/src/stdlib/signal.rs]: We should not allow `signum` to be equal to `NSIG`

File: vm/src/stdlib/signal.rs [hunk 8]
```diff
@@ -81,50 +94,68 @@ fn alarm(time: PyIntRef, _vm: &VirtualMachine) -> u32 {
...
-            let handler = vm
-                .signal_handlers
-                .borrow()
-                .get(&(signum as i32))
-                .expect("Handler should be set")
-                .clone();
-            vm.invoke(&handler, vec![vm.new_int(signum), vm.get_none()])
-                .expect("Test");
+            let handler = &vm.signal_handlers.borrow()[signum];
+            if vm.is_callable(handler) {
```
Comment (palaviv) [inline vm/src/stdlib/signal.rs]: How can the handler be not callable? We check this when we set the signal.

Comment (coolreader18) [reply vm/src/stdlib/signal.rs]: If it's `SIG_DFL` or `SIG_IGN`, that still gets stored in the array and returned as the old handler when you call `signal()`.

File: src/main.rs (hunk 6 above)
Comment (coolreader18) [reply src/main.rs]: Actually no, it doesn't use `code.py`, that's just trying to emulate it. I'll look into this further.

**Comment by coolreader18** [comment]
@windelbouwman I refactored the prompt a bit, and although that's sort of out of scope of the rest of the changes, I feel like it better represents the (data?)flow of the shell.

File: src/main.rs [hunk 9]
```diff
@@ -582,14 +591,26 @@ fn run_shell(vm: &VirtualMachine, scope: Scope) -> PyResult<()> {
     let stdout = io::stdout();
     let mut stdout = stdout.lock();
 
-    print!("{}", get_prompt(vm, "ps1"));
-    stdout.flush().expect("flush failed");
+    let mut print_prompt = || {
```
Comment (windelbouwman) [inline src/main.rs]: This does not need to be `mut`. You might also define a normal function like `fn print_prompt() {` here.

File: vm/src/vm.rs [hunk 10]
```diff
@@ -166,6 +169,8 @@ impl VirtualMachine {
         let import_func = RefCell::new(ctx.none());
         let profile_func = RefCell::new(ctx.none());
         let trace_func = RefCell::new(ctx.none());
+        let signal_handlers = RefCell::new(arr![ctx.none(); 64]);
```
Comment (windelbouwman) [inline vm/src/vm.rs]: This `64` can be the `NSIG` from above right?

Comment (windelbouwman) [reply vm/src/vm.rs]: Ah, okay.

**Pull request merged**